### PR TITLE
Features/rubygem convention

### PR DESCRIPTION
**Pull request by actfong**
Gem to be used to sneak into rails instrumentation channels to perform custom actions.
Will be of great use for collecting metrics data, to be included within Columbus

Please have a look at the README for details

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Copyright (C) 2016 FromAtoB.com http://fromatob.com
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of
+this software and associated documentation files (the "Software"), to deal in
+the Software without restriction, including without limitation the rights to
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
```
Comment (felixlublasser) [inline LICENSE]: Is this standard? Do we really want to give people the right to make money off of the code?

File: README.md [hunk 2]
```diff
@@ -1 +1,186 @@
...
+
+    $ bundle
+
+Or command line:
+
+    $ gem install kenny
+
+## Usage
+  Kenny can be configured through an initializer (`config/initializers/kenny.rb`) or within the configuration file of your environment `development|test|staging|production.rb`.
+  This depends on whether you want to same or different behaviour accross environments.
```
Comment (felixlublasser) [inline README.md]: There is either a word missing or one too much in this sentence :smile: Maybe rephrase.

File: README.md [hunk 3]
```diff
@@ -1 +1,186 @@
...
+Lograge does a great job in suppressing Rails' log output, it does so by:
+- overriding Rails' default behaviour for instrumentations/subscriptions
+- being very opinionated, hence removing a lot of data from your log.
+
+As a result, even though Lograge is great for cleaning your production.log, it is neither meant for collecting metrics nor great for plugging into any instrumentation event that you want.
+
+Kenny attempts to leave implementations to the user and provide a more modular way to monitor instrumentation events.
+It allows the user to decide:
+- which instrumentations to monitor
+- what to do when the specified instrumentation-event occurs
```
Comment (felixlublasser) [inline README.md]: I would write it `instrumentation event`.

File: README.md [hunk 4]
```diff
@@ -1 +1,186 @@
...
+
+  end
+
+  ```
+
+
+### `kenny.instrumentations` configuration
+  Before proceeding, have a look at [Active Support Instrumentation](http://edgeguides.rubyonrails.org/active_support_instrumentation.html) and [LogSubscriber](http://api.rubyonrails.org/classes/ActiveSupport/LogSubscriber.html)
+
+  The `kenny.instrumentations` configuration takes an array of hashes. Each hash represents a Rails instrumentation-event that you want to subscribe to.
```
Comment (felixlublasser) [inline README.md]: See comment above.

File: README.md [hunk 5]
```diff
@@ -1 +1,186 @@
...
+
+  Each of these hashes requires a `:name` (name of instrumentation event), a `:block` (what to do when that event occurs) and *optionally* a `:logger` (which logger to use in order to write these events to a file).
+
+  In the example above, we setup Kenny to monitor two instrumentation events, `process_action.action_controller` and `sql.active_record`.
+
+  For every instrumentation that you want to monitor, you have to provide a `:name` and a `:block`. `:logger` is optional.
```
Comment (felixlublasser) [inline README.md]: You explained this in L81 already, don't think the repetition adds to the understanding.

File: README.md [hunk 6]
```diff
@@ -1 +1,186 @@
...
+
+  Behind the scenes, it defines an anonymous class (< ActiveSupport::LogSubscriber), for each of the specified instrumentation.
```
Comment (felixlublasser) [inline README.md]: `instrumentations`

Comment (felixlublasser) [inline README.md]: Also, no `,` after the parentheses :)

File: README.md [hunk 7]
```diff
@@ -1 +1,186 @@
...
+
+  The first one gets method `def process_action` defined and `def logger` *redefined*.
+
+  The body of `def process_action` is the `:block` that has been supplied to the configuration.
+  Hence `:block` must be a Lambda or a Proc.
```
Comment (felixlublasser) [inline README.md]: `,` after `Hence`

File: README.md [hunk 8]
```diff
@@ -1 +1,186 @@
...
+  (Lambda will raise errors with wrong number of arguments, Proc won't)
+
+  The `def logger` method (in ActiveSupport::LogSubscriber) gets overridden and will return the logger-instance that you have provided to the configuration. (In this case, its the `request_logger` that you have defined at the top of your config).
```
Comment (felixlublasser) [inline README.md]: `logger instance`

Comment (felixlublasser) [inline README.md]: `it's`

File: README.md [hunk 9]
```diff
@@ -1 +1,186 @@
...
+
+  This class is then attached to :action_controller, by looking up the event-name 'process_action.action_controller'
```
Comment (felixlublasser) [inline README.md]: `event name`

File: README.md [hunk 10]
```diff
@@ -1 +1,186 @@
...
+
+  The idea of redefining `def logger` may sound a bit scary, but this is necessary to keep events from
+  different instrumentation channels on different log files. If `:logger` option is not provided, then that LogSubscriber class will use the default Rails logger (and hence writing to your production.log etc)
```
Comment (felixlublasser) [inline README.md]: Either `be writing` or `write`

File: README.md [hunk 11]
```diff
@@ -1 +1,186 @@
...
+
+  The second LogSubscriber class will have method `def start_processing` defined and the method body is again what has been supplied in the :block configuration.
+
+  The difference is that `:logger` has not been provided, hence it won't override the logger method for this LogSubscriber. In Rails, this means it will fall back to the default `Rails.logger`.
+  Mind you that even though a LogSubscriber has access to a logger, it does not mean you have to use it! There is nothing wrong with subscribing to an event and not log its data to file at all! There are other things you could do with this data.
+  At the end, this class gets attached to :active_record, by looking up the event-name 'sql.active_record'
```
Comment (felixlublasser) [inline README.md]: `event name`

File: README.md [hunk 12]
```diff
@@ -1 +1,186 @@
...
+
+#### Be careful with variable scopes and lambdas
+  Since lambda's are used to define method bodies, be careful with context of variables.
```
Comment (felixlublasser) [inline README.md]: `lambdas`

File: README.md [hunk 13]
```diff
@@ -1 +1,186 @@
...
+    { name: 'sql.active_record',
+      block: lambda do |event|
+        data = event.payload
+        logger.info("#{event.name}: #{data}")
+      end
+    }
+  ]
+  ```
+
+  You might think that since no `:logger` option has been provided, for 'sql.active_record' events, the default logger will be used..... But that is not true.
```
Comment (felixlublasser) [inline README.md]: No `,` after `provided`.

Comment (felixlublasser) [inline README.md]: No `,` after `provided`.

File: README.md [hunk 14]
```diff
@@ -1 +1,186 @@
...
+  Since `logger` is within scope at the time when the lambda was defined, this instance of ActiveSupport::Logger will be used to invoke `#info` when 'sql.active_record' occurs. So just avoid creating a local variables that have the same name as variables in your block.
```
Comment (felixlublasser) [inline README.md]: get rid of `a` before `variables`

File: README.md [hunk 15]
```diff
@@ -1 +1,186 @@
...
+
+
+### `kenny.unsubscribe_rails_defaults` configuration
+  Kenny can also used to unsubscribe all Rails LogSubscribers from their subscribed instrumentation-events.
```
Comment (felixlublasser) [inline README.md]: `be used`, `instrumentation events`

File: README.md [hunk 16]
```diff
@@ -1 +1,186 @@
...
+
+  ``` Ruby
+  config.kenny.unsubscribe_rails_defaults = true
+  config.kenny.instrumentations = [{
+      # your stuff
+    }]
+  }
+  ```
+
+  By doing so, your `development|test|staging|production.log` will not have any of the default log messages. This is not an approach I would recommend, unless you are desparate to have all messages from your specified instrumentation-events all logged into one `development|test|staging|production.log`.
```
Comment (felixlublasser) [inline README.md]: `desperate`, `instrumentation events`

File: README.md [hunk 17]
```diff
@@ -1 +1,186 @@
...
+  Started GET "/my_path" for 10.0.2.2 at 2016-07-12 10:06:48 +0000
+  Started GET "/assets/sh/my_styles.css?body=1" for 10.0.2.2 at 2016-07-12 10:06:49 +0000
+  ```
+
+  You can suppress these messages by setting kenny.suppress_rack_logger to true.
+  This setting will not have any effects on the log files that you create separately with the `:logger` option within your specified instrumentation.
+
+## Open-to-Implementation approach
+  As you might have seen from the example, the `:block` allows you to define your own implementation.
+  My idea behind writing this gem, is to free up the user from the tedious task of defining LogSubscriber classes and to allow the user to define whatever (s)he wants to do with the event data, be it something like:
```
Comment (felixlublasser) [inline README.md]: no `,` after `gem`. Maybe get rid of repeated `the user` and use `they` instead. Like this: `[...] and to allow them to define whatever they want to do [...]`.

File: README.md [hunk 18]
```diff
@@ -1 +1,186 @@
...
+          Something.async.process_and_forward(data)
+        end
+      }
+    ]
+  ```
+
+  Again, there is no requirement for you to write messages to log files. It is all up to you.
+
+## Formatters
+  Apart from subscribing to instrumentation-events and logging, Kenny also provides formatters, which you can attach to a logger.
```
Comment (felixlublasser) [inline README.md]: `instrumentation events`

File: README.md [hunk 19]
```diff
@@ -1 +1,186 @@
...
+
+  Currently it only comes with a LogStashFormatter, but feel free to add more Formatters to make this project great.
```
Comment (felixlublasser) [inline README.md]: `,` after `Currently`

File: kenny.gemspec [hunk 20]
```diff
@@ -0,0 +1,44 @@
+# coding: utf-8
+Gem::Specification.new do |spec|
+  spec.name          = "kenny"
+  spec.version       = "0.1.0"
+  spec.authors       = ["Mathias Rüdiger"]
```
Comment (felixlublasser) [inline kenny.gemspec]: No `Alex Fong`? :)

File: lib/kenny.rb [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+  def self.application=(app)
+    @@application = app
+  end
+
+  def self.application
+    @@application
+  end
+
+  ##
+  # Define LogSubscriber-classes and Attach to user-specified instrumentations
```
Comment (felixlublasser) [inline lib/kenny.rb]: `attach` lowercase

File: README.md (hunk 2 above)
Comment (actfong) [inline README.md]: Good one. Noted!

File: lib/kenny.rb [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+    if @@application.config.kenny[:suppress_rack_logger]
+      require "kenny/rails_ext/rack/logger"
+    end
+  end
+
+  ##
+  # Create LogSubscriber-classes which will be attached to the user-specified instrumentations
+  # These classes are anonymous, but inherit from Kenny::LogSubscriber to simplify testing
+  #
+  # Within these classes, methods (and eventually logger) are defined based on the
```
Comment (felixlublasser) [inline lib/kenny.rb]: Why `eventually`? Do you mean `potentially`?

File: README.md (hunk 4 above)
Comment (actfong) [inline README.md]: Noted!

File: LICENSE (hunk 1 above)
Comment (actfong) [inline LICENSE]: Good point. But I suppose that comes with writing Open Source anyway.

Maybe it would be worth for me to go through these licenses at some point:
https://opensource.org/licenses

But for now I will just stick to the (simpler) MIT license. 
I have also added the license from Lograge as one of the classes (Rack::Logger) comes from Lograge.

File: lib/kenny/railtie.rb [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+      # Define anonymous classes that inherit from ActiveSupport::LogSubscriber.
+      # Within that anonymous class, define methods that
+      # perform the user-defined actions when that instrumentation occurs.
+      # If desired, user can define a specific logger for the specified instrumentation.
+      Kenny.attach_to_instrumentations
+
+      # Unsubscribe all default Rails LogSubscribers if demanded
+      Kenny.unsubscribe_from_rails_defaults
+
+      # Suppress Rails Rack::Rack::Logger's output
```
Comment (felixlublasser) [inline lib/kenny/railtie.rb]: `Rails::Rack::Logger`

Comment (actfong) [inline lib/kenny/railtie.rb]: Well spotted
